Timezone not supported +DDDD

Hello, starting from 338 version (it works on 335) we are getting the following error on the query:
```
select timestamp '2019-01-05 23:59:59.999 +0000';
```
`Error: Timezone not supported +0000`
But when changing the query to : `select timestamp '2019-01-05 23:59:59.999 +00:00';`
it works.

```
Query 20200720_201203_00009_qgcxc failed: line 1:8: Time zone not supported: +0000
io.prestosql.spi.PrestoException: line 1:8: Time zone not supported: +0000
	at io.prestosql.sql.analyzer.ExpressionAnalyzer$Visitor.visitTimestampLiteral(ExpressionAnalyzer.java:885)
	at io.prestosql.sql.analyzer.ExpressionAnalyzer$Visitor.visitTimestampLiteral(ExpressionAnalyzer.java:340)
	at io.prestosql.sql.tree.TimestampLiteral.accept(TimestampLiteral.java:52)
	at io.prestosql.sql.tree.StackableAstVisitor.process(StackableAstVisitor.java:27)
	at io.prestosql.sql.analyzer.ExpressionAnalyzer$Visitor.process(ExpressionAnalyzer.java:363)
	at io.prestosql.sql.analyzer.ExpressionAnalyzer.analyze(ExpressionAnalyzer.java:301)
	at io.prestosql.sql.analyzer.ExpressionAnalyzer.analyzeExpression(ExpressionAnalyzer.java:1693)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.analyzeExpression(StatementAnalyzer.java:2361)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.analyzeSelectSingleColumn(StatementAnalyzer.java:2201)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.analyzeSelect(StatementAnalyzer.java:2025)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:1285)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:303)
	at io.prestosql.sql.tree.QuerySpecification.accept(QuerySpecification.java:144)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:320)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:330)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:915)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:303)
	at io.prestosql.sql.tree.Query.accept(Query.java:107)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:320)
	at io.prestosql.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:289)
	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:83)
	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:75)
	at io.prestosql.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:246)
	at io.prestosql.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:180)
	at io.prestosql.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:746)
	at io.prestosql.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:121)
	at io.prestosql.$gen.Presto_unknown____20200720_183506_2.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.prestosql.spi.type.TimeZoneNotSupportedException: Time zone not supported: +0000
	at io.prestosql.spi.type.TimeZoneKey.getTimeZoneKey(TimeZoneKey.java:133)
	at io.prestosql.spi.type.DateTimeEncoding.packDateTimeWithZone(DateTimeEncoding.java:38)
	at io.prestosql.type.Timestamps.parseShortTimestampWithTimeZone(Timestamps.java:441)
	at io.prestosql.type.Timestamps.parseTimestampWithTimeZone(Timestamps.java:340)
	at io.prestosql.sql.analyzer.ExpressionAnalyzer$Visitor.visitTimestampLiteral(ExpressionAnalyzer.java:871)
	... 34 more
```

--- a/presto-spi/src/main/java/io/prestosql/spi/type/TimeZoneKey.java
+++ b/presto-spi/src/main/java/io/prestosql/spi/type/TimeZoneKey.java
@@ -23,6 +23,7 @@ import java.time.ZoneId;
 import java.util.Collections;
 import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Objects;
@@ -31,7 +32,6 @@ import java.util.Set;
 import java.util.TreeMap;
 
 import static io.prestosql.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
-import static java.lang.Character.isDigit;
 import static java.lang.Math.abs;
 import static java.lang.Math.max;
 import static java.lang.String.format;
@@ -50,6 +50,47 @@ public final class TimeZoneKey
     private static final short OFFSET_TIME_ZONE_MIN = -14 * 60;
     private static final short OFFSET_TIME_ZONE_MAX = 14 * 60;
     private static final TimeZoneKey[] OFFSET_TIME_ZONE_KEYS = new TimeZoneKey[OFFSET_TIME_ZONE_MAX - OFFSET_TIME_ZONE_MIN + 1];
+
+    private static final Set<String> UTC_EQUIVALENTS = Set.of(
+            "gmt",
+            "gmt0",
+            "gmt+0",
+            "gmt-0",
+            "etc/gmt",
+            "etc/gmt0",
+            "etc/gmt+0",
+            "etc/gmt-0",
+            "ut",
+            "ut+0",
+            "ut-0",
+            "etc/ut",
+            "etc/ut+0",
+            "etc/ut-0",
+            "utc",
+            "utc+0",
+            "utc-0",
+            "etc/utc",
+            "etc/utc+0",
+            "etc/utc-0",
+            "+0000",
+            "+00:00",
+            "-0000",
+            "-00:00",
+            "z",
+            "zulu",
+            "uct",
+            "greenwich",
+            "universal",
+            "etc/universal",
+            "etc/uct");
+
+    private static final List<String> UTC_OFFSET_PREFIXES = List.of(
+            "etc/gmt",
+            "etc/utc",
+            "etc/ut",
+            "gmt",
+            "utc",
+            "ut");
 
     static {
         try (InputStream in = TimeZoneKey.class.getResourceAsStream("zone-index.properties")) {
@@ -210,102 +251,95 @@ public final class TimeZoneKey
         return normalizeZoneId(zoneId).equals("utc");
     }
 
-    private static String normalizeZoneId(String originalZoneId)
+    private static String normalizeZoneId(String zoneId)
     {
-        String zoneId = originalZoneId.toLowerCase(ENGLISH);
+        if (isUtcEquivalentName(zoneId)) {
+            return "utc";
+        }
 
-        boolean startsWithEtc = zoneId.startsWith("etc/");
-        if (startsWithEtc) {
-            zoneId = zoneId.substring(4);
+        // The JDK doesn't understand legacy timezones such as Etc/GMT+00:00, GMT-7
+        String lowerCased = zoneId.toLowerCase(ENGLISH);
+        for (String prefix : UTC_OFFSET_PREFIXES) {
+            if (lowerCased.startsWith(prefix)) {
+                int offset = prefix.length();
+                if (offset == lowerCased.length() || lowerCased.charAt(offset) != '+' && lowerCased.charAt(offset) != '-') {
+                    // It must be something like GMT7, or Etc/UTC4, which are invalid. Unsupported timezones are handled by the caller.
+                    return zoneId;
+                }
+
+                int colon = lowerCased.indexOf(':', offset);
+                int hour;
+                int minute = 0;
+                try {
+                    if (colon != -1) {
+                        hour = Integer.parseInt(lowerCased, offset, colon, 10);
+                        minute = Integer.parseInt(lowerCased, colon + 1, lowerCased.length(), 10);
+                    }
+                    else if (lowerCased.length() - offset <= 3) {
+                        // +H or +HH
+                        hour = Integer.parseInt(lowerCased, offset, lowerCased.length(), 10);
+                    }
+                    else {
+                        // let the JDK handle it below
+                        break;
+                    }
+                }
+                catch (NumberFormatException e) {
+                    // Invalid time zone is handled by the caller
+                    return zoneId;
+                }
+
+                if (hour == 0 && minute == 0) {
+                    return "utc";
+                }
+
+                if (prefix.equals("etc/gmt")) {
+                    hour = -hour;
+                }
+
+                return formatZoneOffset(hour, minute);
+            }
+        }
+
+        // Normalize using JDK rules. In particular, normalize forms such as +0735, +7, +073521
+        try {
+            zoneId = ZoneId.of(zoneId).getId();
+        }
+        catch (Exception e) {
+            return zoneId;
         }
 
         if (isUtcEquivalentName(zoneId)) {
             return "utc";
         }
 
-        //
-        // Normalize fixed offset time zones.
-        //
+        return zoneId;
+    }
 
-        // In some zones systems, these will start with UTC, GMT or UT.
-        int length = zoneId.length();
-        boolean startsWithEtcGmt = false;
-        if (length > 3 && (zoneId.startsWith("utc") || zoneId.startsWith("gmt"))) {
-            if (startsWithEtc && zoneId.startsWith("gmt")) {
-                startsWithEtcGmt = true;
-            }
-            zoneId = zoneId.substring(3);
-            length = zoneId.length();
-        }
-        else if (length > 2 && zoneId.startsWith("ut")) {
-            zoneId = zoneId.substring(2);
-            length = zoneId.length();
-        }
+    private static String formatZoneOffset(int hour, int minute)
+    {
+        StringBuilder builder = new StringBuilder();
 
-        // (+/-)00:00 is UTC
-        if ("+00:00".equals(zoneId) || "-00:00".equals(zoneId)) {
-            return "utc";
-        }
+        builder.append(hour >= 0 ? '+' : '-');
 
-        // if zoneId matches XXX:XX, it is likely +HH:mm, so just return it
-        // since only offset time zones will contain a `:` character
-        if (length == 6 && zoneId.charAt(3) == ':') {
-            return zoneId;
+        hour = abs(hour);
+        if (hour < 10) {
+            builder.append('0');
         }
+        builder.append(hour);
+        builder.append(':');
 
-        //
-        // Rewrite (+/-)H[H] to (+/-)HH:00
-        //
-        if (length != 2 && length != 3) {
-            return originalZoneId;
+        if (minute < 10) {
+            builder.append('0');
         }
+        builder.append(minute);
 
-        // zone must start with a plus or minus sign
-        char signChar = zoneId.charAt(0);
-        if (signChar != '+' && signChar != '-') {
-            return originalZoneId;
-        }
-        if (startsWithEtcGmt) {
-            // Flip sign for Etc/GMT(+/-)H[H]
-            signChar = signChar == '-' ? '+' : '-';
-        }
-
-        // extract the tens and ones characters for the hour
-        char hourTens;
-        char hourOnes;
-        if (length == 2) {
-            hourTens = '0';
-            hourOnes = zoneId.charAt(1);
-        }
-        else {
-            hourTens = zoneId.charAt(1);
-            hourOnes = zoneId.charAt(2);
-        }
-
-        // do we have a valid hours offset time zone?
-        if (!isDigit(hourTens) || !isDigit(hourOnes)) {
-            return originalZoneId;
-        }
-
-        // is this offset 0 (e.g., UTC)?
-        if (hourTens == '0' && hourOnes == '0') {
-            return "utc";
-        }
-
-        return "" + signChar + hourTens + hourOnes + ":00";
+        return builder.toString();
     }
 
     private static boolean isUtcEquivalentName(String zoneId)
     {
-        return zoneId.equals("utc") ||
-                zoneId.equals("z") ||
-                zoneId.equals("ut") ||
-                zoneId.equals("uct") ||
-                zoneId.equals("gmt") ||
-                zoneId.equals("gmt0") ||
-                zoneId.equals("greenwich") ||
-                zoneId.equals("universal") ||
-                zoneId.equals("zulu");
+        return UTC_EQUIVALENTS.contains(zoneId.toLowerCase(ENGLISH));
     }
 
     private static String zoneIdForOffset(long offset)

--- a/presto-spi/src/test/java/io/prestosql/spi/type/TestTimeZoneKey.java
+++ b/presto-spi/src/test/java/io/prestosql/spi/type/TestTimeZoneKey.java
@@ -61,6 +61,8 @@ public class TestTimeZoneKey
         assertSame(TimeZoneKey.getTimeZoneKey("GMT-00:00"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("+00:00"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("-00:00"), UTC_KEY);
+        assertSame(TimeZoneKey.getTimeZoneKey("+0000"), UTC_KEY);
+        assertSame(TimeZoneKey.getTimeZoneKey("-0000"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("etc/utc"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt"), UTC_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("etc/gmt+0"), UTC_KEY);
@@ -140,6 +142,19 @@ public class TestTimeZoneKey
         assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC-7"), MINUS_7_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC+7"), PLUS_7_KEY);
         assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC-7"), MINUS_7_KEY);
+
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC-7:00"), MINUS_7_KEY);
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC-07:00"), MINUS_7_KEY);
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC+7:00"), PLUS_7_KEY);
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC+07:00"), PLUS_7_KEY);
+
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC-7:35"), TimeZoneKey.getTimeZoneKeyForOffset(-(7 * 60 + 35)));
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC-07:35"), TimeZoneKey.getTimeZoneKeyForOffset(-(7 * 60 + 35)));
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC+7:35"), TimeZoneKey.getTimeZoneKeyForOffset(7 * 60 + 35));
+        assertSame(TimeZoneKey.getTimeZoneKey("Etc/UTC+07:35"), TimeZoneKey.getTimeZoneKeyForOffset(7 * 60 + 35));
+
+        assertSame(TimeZoneKey.getTimeZoneKey("+0735"), TimeZoneKey.getTimeZoneKeyForOffset(7 * 60 + 35));
+        assertSame(TimeZoneKey.getTimeZoneKey("-0735"), TimeZoneKey.getTimeZoneKeyForOffset(-(7 * 60 + 35)));
     }
 
     @Test

